draco/fact_utils.py

```diff
@@ -78,7 +80,134 @@ def dict_to_facts(
+    return result
+
+
+def handle_path(
+    address: List[Union[str, int]], value: Any, nested_dict: Dict, kind: FactKind
+) -> Dict[Union[str, int], Any]:
+    assert len(address) != 0
+    if len(address) == 1:
+        last = address[0]
+        if kind == FactKind.ATTRIBUTE.value:
+            nested_dict[last] = value
+        else:
+            if last in nested_dict:
+                nested_dict[last].append(value)
+            else:
+                nested_dict[last] = [value]
+        return nested_dict
+
+    else:
+        elem = address[0]
+        if elem in nested_dict:
+            nested_dict[elem] = handle_path(address[1:], value, nested_dict[elem], kind)
+        else:
+            nested_dict[elem] = handle_path(address[1:], value, dict(), kind)
+        return nested_dict
+
+
 def facts_to_dict(facts: List) -> Mapping:
-    """A generic decoder that converts an answer set into a nested data structure"""
-    pass  # TODO: https://github.com/cmudig/draco2/issues/24
-    return {}
+    """
+    A generic decoder that converts an answer set into a nested data structure
+
+    """
+    Root: Dict[Union[str, int], Any] = dict()
+    # Creating the memo dictionary that maps numbers to nested properties
+    for fact in facts:
+        assert fact.type == SymbolType.Function
+        kind, values = fact.name, fact.arguments
+        parsed_values = parse_values(values)
+        (key, parent_index, val) = parsed_values
+        if kind == FactKind.PROPERTY.value:
+            if parent_index == ROOT:
+                if type(key) == tuple:
+                    Root = handle_path(list(key), val, Root, kind)
+                else:
+                    if key in Root:
+                        Root[key].append(val)
+                    else:
+                        Root[key] = [val]
+            else:
+                if type(key) == tuple:
+                    Root[parent_index] = handle_path(
+                        list(key), val, Root[parent_index], kind
+                    )
+                else:
+                    if key in Root[parent_index]:
+                        Root[parent_index][key].append(val)
+                    else:
+                        Root[parent_index][key] = [val]
+
+            if val not in Root:
+                Root[val] = dict()
+        else:
+            if parent_index == ROOT:
+                if type(key) == tuple:
+                    Root = handle_path(list(key), val, Root, kind)
+                else:
+                    Root[key] = val
+            else:
+                if parent_index not in Root:
+                    Root[parent_index] = dict()
+                if type(key) == tuple:
+                    Root[parent_index] = handle_path(
+                        list(key), val, Root[parent_index], kind
+                    )
+                else:
+                    Root[parent_index][key] = val
+    # Running through the memoized dictionary to replace the integer ID
```

Comment:
remove comments that say what the code does and add comments that explain why the code does things